### PR TITLE
chore(backend): Consolidate v2 placeholder resolution into shared utility. Fixes #12949

### LINKED ISSUE #12949: chore(backend): Consolidate v2 placeholder resolution into shared utility
**TL;DR:** Driver and launcher currently implement overlapping placeholder/type-resolution logic. Consolidate shared behavior into one reusable utility to reduce drift and keep future fixes/tests in a single path.

## Chore description

Follow up on PR #12598 by consolidating placeholder resolution logic used in:

- `backend/src/v2/driver/` (driver-side argument/command preprocessing)
- `backend/src/v2/component/` (launcher-side command/arg compilation)

The goal is to extract shared logic (input placeholder parsing + typed value stringification) into a common, reusable package and keep layer-specific behavior in thin wrappers.

## Why is this needed?

There is pre-existing overlap between driver and launcher placeholder resolution, and PR #12598 expanded that overlap. This increases maintenance risk:

- the same placeholder/type bug can require fixes in multiple places
- behavior may diverge subtly over time (formatting, null handling, edge cases)
- test coverage can drift by layer instead of validating one canonical behavior

This is non-blocking for PR #12598, but should be addressed as follow-up hardening.

### Labels

/area backend

---

## Proposed solution

1. Add a shared utility in a neutral/common backend package for:
   - resolving `{{$.inputs.parameters['...']}}` placeholders in strings (including embedded/multiple placeholders)
   - canonical conversion from `structpb.Value` to string for supported kinds

2. Keep layer-specific logic outside shared utility:
   - `IfPresent` expansion policy stays in driver wrapper
   - output/artifact placeholder mapping stays in launcher wrapper
   - command-vs-args validation stays at call sites

3. Update driver and launcher to call the shared utility for common behavior instead of duplicating it.

4. Add parity tests that assert equivalent results across driver/launcher for shared scenarios.

---

### In scope

- Refactor to remove duplicate input-placeholder/type-resolution logic across driver and launcher
- Keep runtime behavior unchanged for existing passing scenarios
- Add regression tests for shared behavior

### Out of scope

- Changing overall PR #12598 feature semantics (`IfPresent` behavior)
- Broad refactors unrelated to placeholder resolution
- Changing artifact/output placeholder contracts

---

## Acceptance criteria

- [ ] Shared utility exists in common backend package and is used by both driver and launcher for input-placeholder/type-resolution behavior.
- [ ] No duplicate typed-stringification implementations remain for shared placeholder paths.
- [ ] Driver and launcher behavior matches for shared placeholder scenarios (string/number/bool/list/struct/null where applicable).
- [ ] Regression tests added for:
  - [ ] embedded placeholder substitution
  - [ ] multiple placeholders in one token
  - [ ] typed values conversion parity
  - [ ] missing-parameter error behavior parity
- [ ] Existing tests in touched packages remain green.

---

Love this idea? Give it a 👍.

### PR DESCRIPTION
**Pull request by khushiiagrawal**
**Description of your changes:**

This PR consolidates duplicate placeholder resolution and type stringification logic that existed in both the driver and launcher into a shared utility package, addressing technical debt from PR #12598.

Fixes #12949

**What changed:**
- Created new `backend/src/v2/placeholder/` package with:
  - `PbValueToString()` - canonical `structpb.Value` to string conversion
  - `ResolveInputParameterPlaceholders()` - shared input parameter placeholder resolution
  - `InputParameterRe` - exported regex pattern for placeholder matching

- Updated `backend/src/v2/driver/util.go`:
  - Removed duplicate `pbValueToString`, `resolveSinglePlaceholder`, `resolveInputParameterPlaceholders` functions
  - Now calls shared package for common logic
  - Kept driver-specific logic (IfPresent expansion, output placeholder handling)

- Updated `backend/src/v2/component/launcher_v2.go`:
  - Replaced 25-line inline type conversion switch with single call to shared function
  - Fixed latent bug where null values weren't handled

**Why:**
- Eliminates code duplication between driver and launcher
- Prevents behavior drift across layers
- Centralizes type conversion and placeholder resolution for easier maintenance
- All existing tests pass with no behavioral changes

**Checklist:**
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by khushiiagrawal** [comment]
@droctothorpe  @hbelmiro  PTAL. thanks.

**Comment by hbelmiro** [comment]
@khushiiagrawal can you please rebase?

**Comment by khushiiagrawal** [comment]
@hbelmiro  could you please take a look! 
thanks

File: backend/src/common/util/workflow_test.go
Comment (hbelmiro) [inline backend/src/common/util/workflow_test.go:1]: These changes are unrelated.

File: backend/src/v2/placeholder/placeholder.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	switch t := v.Kind.(type) {
+	case *structpb.Value_NullValue:
+		return "", nil
+	case *structpb.Value_StringValue:
+		return v.GetStringValue(), nil
+	case *structpb.Value_NumberValue:
+		return strconv.FormatFloat(v.GetNumberValue(), 'f', -1, 64), nil
+	case *structpb.Value_BoolValue:
+		return strconv.FormatBool(v.GetBoolValue()), nil
```
Comment (hbelmiro) [inline backend/src/v2/placeholder/placeholder.go:57]: **Silent behavioral change in number formatting for the driver path.** The old driver used `fmt.Sprintf("%g", n)` for non-integer floats, which produces scientific notation for extreme values (e.g., `1e+08`). The new shared code uses `strconv.FormatFloat(v, 'f', -1, 64)`, which always produces decimal notation (e.g., `100000000`). This changes container arguments resolved through `resolveContainerArgs` and `resolveCondition`.

**Suggested fix:** Preserve the old driver formatting in `PbValueToString` so this refactor is purely structural with no behavioral change.

**Comment by khushiiagrawal** [comment]
@hbelmiro  thanks for the review, i have addressed the changes!

File: backend/src/v2/placeholder/placeholder.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// InputParameterRe matches the complete {{$.inputs.parameters['name']}}
+// placeholder including the surrounding double braces.
+var InputParameterRe = regexp.MustCompile(`\{\{\$\.inputs\.parameters\['(.+?)']}}`)
```
Comment (hbelmiro) [inline backend/src/v2/placeholder/placeholder.go]: Please make it unexported.

```suggestion
var inputParameterRe = regexp.MustCompile(`\{\{\$\.inputs\.parameters\['(.+?)']}}`)
```

**Review by hbelmiro — COMMENTED** [review]
Left a minor comment. Also, can you please rebase?

**Comment by khushiiagrawal** [comment]
@hbelmiro  PTAL, thanks!

**Review by hbelmiro — APPROVED** [review]
/lgtm
/approve
/ok-to-test

**Comment by hbelmiro** [comment]
@khushiiagrawal can you please check the CI?

**Comment by khushiiagrawal** [comment]
> @khushiiagrawal can you please check the CI?

Its done! @hbelmiro , PTAL

**Comment by khushiiagrawal** [comment]
@hbelmiro  PTAL one you get a chance, thanks!

File: backend/src/v2/placeholder/placeholder.go (hunk 2 above)
Comment (hbelmiro) [inline backend/src/v2/placeholder/placeholder.go]: Still exported.

File: backend/src/v2/placeholder/placeholder.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	case *structpb.Value_StringValue:
+		return v.GetStringValue(), nil
+	case *structpb.Value_NumberValue:
+		n := v.GetNumberValue()
+		if n == float64(int64(n)) {
+			return fmt.Sprintf("%d", int64(n)), nil
+		}
+		return fmt.Sprintf("%g", n), nil
+	case *structpb.Value_BoolValue:
+		return strconv.FormatBool(v.GetBoolValue()), nil
```
Comment (hbelmiro) [inline backend/src/v2/placeholder/placeholder.go:57]: The current code uses `fmt.Sprintf("%g", n)` which switches to scientific notation for small floats (exponent < -4). The old launcher used `strconv.FormatFloat(n, 'f', -1, 64)` which always produces decimal notation. Decimal is more predictable for ML parameters like learning rates (`0.00001` vs `1e-05`).

Please change line 57 to use decimal notation so both driver and launcher unify on `'f'` format:

```go
return strconv.FormatFloat(n, 'f', -1, 64), nil
```

**Review by hbelmiro — COMMENTED** [review]
/approve cancel

**Comment by khushiiagrawal** [comment]
@hbelmiro  I have addressed the reviews, PTAL.
Thanks!

**Comment by khushiiagrawal** [comment]
@hbelmiro  a gentle ping ! :)
